### PR TITLE
[Post Search] Add Grouped Tag Searches

### PR DESCRIPTION
**Pull request by jmortiger**
Another attempt at #559, #625, & forum topics [26167](https://e621.net/forum_topics/26167), [28423](https://e621.net/forum_topics/28423), [29607](https://e621.net/forum_topics/29607), [30635](https://e621.net/forum_topics/30635), [33742](https://e621.net/forum_topics/33742), [41547](https://e621.net/forum_topics/41547), [43040](https://e621.net/forum_topics/43040), [43480](https://e621.net/forum_topics/43480?page=1#forum_post_402247), [42551](https://e621.net/forum_topics/42551), [45756](https://e621.net/forum_topics/45756),[54747](https://e621.net/forum_topics/54747), & [56006](https://e621.net/forum_topics/56006).

This adds the ability for users to group search queries together (e.g. `~( female duo ) ~( male solo )` would retrieve posts tagged with both `female` & `duo` or tagged with both `male` & `solo`).

Similarly to #625, while I've years of programming experience in multiple languages, this is my first time developing with Ruby. I've tested this a fair amount, and added whatever relevant tests I can think of, but I'd like to have someone check my work to make sure I've properly accounted for all the consequences of this change.

## Syntax
To create a grouping, surround the grouped terms with a parenthesis & a space (e.g. `( ` & ` )`). This grouping uses the same syntax as any other solo tag; it can be negated by prefixing with `-` (e.g. `-( female duo )`) & made an inclusive or by prefixing with `~` (e.g. `~solo ~( female )`). 

Essentially, all prior syntactic grammar applies, treating groups as an independent search modifying the parent search based on the prefix (or lack thereof). 

#### `(` & `)`
As implemented, the opening and closing parentheses (`(` & `)`) are technically still searchable as tags (unbalanced group delimiters are ignored by the tokenizer and are processed as normal tags), but, considering the almost no posts that use them (and their questionable utility in those cases), it would likely be wise to remove these as valid tags anyway, especially since they are both invalid tags at time of writing. At time of writing, [`(`](https://e621.net/tags?commit=Search&search%5Bhide_empty%5D=0&search%5Bname_matches%5D=%28) is an invalid tag with [0](https://e621.net/posts?tags=%28) posts, and [`)`](https://e621.net/tags?commit=Search&search%5Bhide_empty%5D=0&search%5Bname_matches%5D=%29) is an invalid tag with [1](https://e621.net/posts?tags=%29) post.

### Tokenizer
Instead of manually removing quoted metatags and separating the remainder by whitespace, a regular expression is used. This regex can be tested [here](https://regex101.com/r/KfFYz4). V10 preserved [some quirks of the older tokenizer](https://github.com/e621ng/e621ng/pull/819#issuecomment-2718483519), the latest version is the currently implemented one.

### Main Functionality
This works by making searches recursive, as detailed [here](https://github.com/e621ng/e621ng/issues/559#issuecomment-1874651856). The max depth of this recursion is limited through a constant on `TagQuery`. Searches scan all levels of the search to properly count tokens and to hoist global metatags (e.g. `order`, `limit`, & `randseed`) to the top level, but leave the rest of the group body as the interior sub-search string in the `groups` property on `TagQuery` instances. The query builders take a query string and generate a `TagQuery` instance from it, proceeding as normal. Before concluding the search, the search calls `add_group_search_relation`, which creates a new instance of the query builder with that group's contents and properly modifies the resulting query.

### `status:deleted` Filtering
In order to make `status:deleted` filtering consistent and predictable, any subsearch/level containing the metatags `delreason`, `deletedby`, or `status` with a value of `deleted`, `active`, `any`, or `all` turns off automatic filtering. If `ElasticPostQueryBuilder::GLOBAL_DELETED_FILTER` is set to true, any subsearch/level turning off automatic filtering propagates that state to all other subsearches/levels, including siblings. If false, this is only propagated to ancestors and descendants, not to siblings.

### Group Searching with SQL queries (e.g. comment search, note search, upload search, ...)
Currently half implemented, I'd probably push it off to a later PR. The way Rails combines `ActiveRecord::Relation` objects (`.or` & `.and`) doesn't add the necessary parenthesis to group sub-expressions. I could use `Arel` to construct it, but then I'd need to change the callers to handle that. If I get to it, I get to it, but, as is, changing the constant `PostQueryBuilder::CAN_HAVE_GROUPS` to `false` will disable this entirely, preserving the current behavior.

## Differences from prior attempts
The biggest difference from #625 is that this implementation uses the same syntax logic; the only syntax difference is the addition of the groups, with groups behaving identically to a single token. Although there is no explicit syntax to add a "negated or" (e.g. `~a ~-b` for "a OR NOT b"), groups innately enable this behavior (e.g. `~a ~( -b )` for "a OR NOT b"). As a result of this, almost none of the existing parsing logic is changed; everything is built on top of it to handle groups. This holds true for things like client-side tag suggestions, related tags, etc.

## Itemized changes
Largely different:
- app/logical/tag_query.rb (overhauled tokenizing & added a store for groups used by `ElasticPostQueryBuilder`)
- app/logical/elastic_post_query_builder.rb (added a recursive step for groups, changed insertion of ` -status:deleted` as described above)

Slightly Tweaked:
- app/logical/posts_sets/post.rb (use new/modified `TagQuery` methods for determining insertion of ` -status:deleted`, properly tokenizing recursive searches for creating (`public_`)`tag_array` & (`public_`)`tag_string`, & finding `limit`)
- app/logical/post_query_builder.rb (added optional recursive step for groups; changes to this file can be discarded and `can_have_groups: false` added to the instantiation at the top of `PostQueryBuilder.search` if this shouldn't be enabled for SQL searches without impact anything else)
- app/logical/stats_updater.rb (use preexisting `always_show_deleted` property where appropriate in tag searches)
- app/models/post.rb (enabled calls to `has_tag?` & `fetch_tag` to optionally toggle on recursion)
- app/logical/parse_value.rb (tweaked to eliminate a bug)

Functionally Identical:
- app/logical/elastic_query_builder.rb (refactored existing logic to facilitate recursion)

Tests:
- test/unit/tag_query_test.rb: Added tests for new parsing/tokenizing methods & instantiation, changed test to account for tokenizer now maintaining token order
- test/unit/elastic_post_query_builder_test.rb: Added tests for proper insertion of `-status:deleted`
- test/unit/post_test.rb: Added tests for groups; changed `status` test to reflect new behavior
- test/unit/post_sets/post_test.rb: added a test to ensure `limit` is correctly extracted from grouped query
- test/unit/comment_test.rb: Added test to validate searching post tags through `Post::tag_match_sql` & a conditional test for group searching.

Other Files
- .rubocop_todo.yml: Self-explanatory
- Gemfile & Gemfile.lock: Additions for testing, can be removed with no impact
- docker-compose.yml: Added services to enable running Rails commands; was used to run Benchmark.ips for performance testing; can be removed with no impact

#### Metatags (e.g. `status:modqueue`) & Quoted Metatags (e.g. `description:"text here"`)
* Quoted metatags aren't distinguished from non-quoted metatags nor standard tags in `TagQuery.parse_query`; to ensure consistency, even malformed quoted metatags have leading and trailing double quotes removed.
* All proper quoted metatags are now required to be separated from following tags by whitespace; they will otherwise be treated as non-quoted metatags.
* In order to simplify & optimize in multiple locations, metatags now require some contents, including after quote removal
#### `has_groups?` Implementation
* I think the one I chose works just fine, but if there's dissatisfaction, or someone wants to improve it later, here's 2 gists to help; [my test runner](https://gist.github.com/jmortiger/19d7930a5c3c5df998ca0d5280269129) using [BenchmarkIPS](https://github.com/evanphx/benchmark-ips) & [a version of `TagQuery` that works with it](https://gist.github.com/jmortiger/ccffeb926785793e32a91f47bf6b15e3). This version has a bunch of alternative versions already there. Changing the value of `HAS_GROUPS_VERSIONS` changes which of the corresponding methods are tested. Assuming you have a `rails` service in your docker-compose.yml, you can run the benchmark with this command:
```
docker compose run --rm rails r test/performance/unit/tag_query_has_groups_test.rb
```
   * Since it's not an instance method, tweaking this to pull alternate implementations into the same file as the test runner and injecting those dependencies instead of passing a symbol might be wise.

### CODE REVIEW

**Comment by Tarrgon** [comment]
I wish you luck.

**Comment by jmortiger** [comment]
I don't think there's much more I can do to make this more solid & robust. Excluding keeping this up to date & *maybe* adding more tests (if I can think of any), I'll likely leave as is for now, if anyone's up to looking it over.

**Comment by Sindrake** [comment]
I apologize for not chiming in earlier.
I did not realize just how much work you've been putting into this.

I've looked over the PR, and I don't see any immediate issues.
However, my concern remains the same as with the previous attempts to implement this, which is whether we even want this kind of functionality in the first place. I do not have the data on hand, but if I remember correctly the complexity of searches in general is fairly low. They usually only include only a couple of tags – oftentimes just one.

So this new search format would be used almost exclusively by experienced users.
At the same time, it increases complexity and might impact performance for all users. I would like to know if you've done any kind of performance testing here.

I am happy that the group syntax in this PR is straightforward.
That was one of my notable complaints regarding Tarrgon's implementation, actually.

Either way, the process of reviewing and testing this PR would be challenging for me, since my time is quite limited.
Before that can happen, I would like to survey the community to see if kind of thing is actually widely desired.

**Comment by liamwhite** [comment]
> might impact performance for all users

I am the [original author](https://github.com/e621ng/e621ng/pull/46) of the search support in this codebase.

The performance impact of a `bool` query's `should` clause is identical to that of a `must` when wrapped in a `constant_score` parent. This is because the underlying operation on each shard of each search node must process the same number of indices in the bitmaps to be combined, whether the disjunction or conjunction of two terms is computed.

For both disjunction and conjunction, the time complexity of executing the search is multiplicative O(mn), where m is the number of clauses, and n is the total number of documents in the index (not the total which will be retrieved by the query). m is already limited to 40, and n's impact on the actual time it takes to compute the result set is scaled by a very low constant factor.

It is notable that both `must` and `should` clauses within a `bool` query _are_ more expensive when not wrapped in a `constant_score` query, but the extent to which this is true when sorting by a natural field (instead of `_score`) is so small that it is probably not worth caring about (scores only have to be computed for the top docs returned to the collector, vs for every document in the index when sorting by `_score`).

I also was previously involved with a separate image booru that supports OR clauses in its [search](https://derpibooru.org/pages/search_syntax#or-expressions), and also contains at a fairly [large number](https://derpibooru.org/pages/stats) of indexed documents, and I can say that supporting OR clauses is a total non-issue.

**Comment by jmortiger** [comment]
> I apologize for not chiming in earlier.

No worries, gave me time to put things in the best state anyways.

> I would like to know if you've done any kind of performance testing here.

Aside from vague observations about how long tests took to complete, I haven't; considering my relative inexperience, I didn't want to touch the config/gems to find a performance toolkit, although I did make a few configurable constants in `ElasticPostQueryBuilder` & `TagQuery` that can be changed to improve performance/change desired behavior. From what I can tell without having done any profiling, there's 3 potential new sources of (computational) performance cost; tokenizing, recursive parsing, and the actual search operation on a nested query.

#### Recursive parsing
With the exception of global tag hoisting in `TagQuery::scan_search` & looking for metatags that override the automatic `-status:deleted` injection.
* For `status:deleted`: If `ElasticPostQueryBuilder::GLOBAL_DELETED_FILTER` is true, that would require only a single check at the root of the search to determine that for the entire tree, and that can be passed in to its recursive children to bypass that check (& bypass reinserting that requirement; unless filtering out deleted posts at each leaf node is revealed to be more performant than doing so as high up in the search as possible, which seems unlikely based on @liamwhite's comments).
* For metatag hoisting: this should already be done once ~, and I'm a little irritated I didn't think about that before~; emptying the tags to be hoisted in the parameters to the children will bypass this automatically.

#### Performing the search
Aside from @liamwhite's the aforementioned usage of `constant_score` (which currently isn't in use, but I'll look into) & the potential benefit of global deleted filtering, there are only two possible ways to improve this that I can see.
##### Reducing the number and/or depth of recursive searches
Currently, the recursive depth is tracked & capped at 10 (i.e. nesting 9 groups within each other + the 1 top level ungrouped search is allowed, any more isn't), but that was plucked out of a hat; all of the tests & code are tied to the constant `TagQuery::DEPTH_LIMIT`, so simply changing that to a smaller number (5 is all I'd imagine being necessary) would reduce the maximum. If the number need to be tracked, that can be added and done in a similar way to the tag count. On that note;
##### Reducing the number of overall tags (for groups specifically)
Although increasing the tag limit from 40 was floated in #625, I'd argue it's possible verbose queries might require less tags with groups. I'd imagine most large queries are either
* Casting a wide net w/ many optional tags & then slimming it down w/ negative tags (which would require less tags if you could precisely negate specific tags); or
* Searching with a high level of specificity (which might be improved or hindered with a stricter limit, but only applying the stricter limit to groups would at least make it a lateral move if not a straight improvement).

#### Tokenization
I'd imagine it's largely a non-issue, but I can tweak the tokenizer (I used a different method before I switched to that monstrous regex; I can polish it up and make using it a constant flag like `ElasticPostQueryBuilder::GLOBAL_DELETED_FILTER` if it would help). Most all tokenization is either memoized or only done once as far as I can tell.

> It is notable that both must and should clauses within a bool query are more expensive when not wrapped in a constant_score query 

The current code (inside and outside this PR) doesn't use `constant_score` at all. I'll see if there's an uplift using that.

**Comment by liamwhite** [comment]
> I'll see if there's an uplift using that.

Unless you are benchmarking with data from millions of posts, it will probably be beyond your measuring error.

**Comment by jmortiger** [comment]
Thanks for the heads up; I'll focus on everything leading up to the query.

> At the same time, it increases complexity and might impact performance for all users.

I realize I never explicitly said it before, so let me correct that; with the potential exception of the tokenizing regex (which shouldn't actually matter, as it'll only attempt to match a group if it starts with a `( `, [which is unlikely](https://e621.net/tags?commit=Search&search%5Bhide_empty%5D=0&search%5Bname_matches%5D=%28&search%5Border%5D=date)), the implementation I've used *should* only impact performance for queries that include a group (though I obviously need to test this to be sure).

**Comment by jmortiger** [comment]
As of the most recent changes, non-grouped searches are almost identical in performance, with the grouped implementation performing slightly better on runs against  the same ~1000 post database. That being said, I'm pretty inexperienced with profiling and I'm not sure what discrepancies are statistically relevant. When I have the time to validate and properly format this, I'll add that (if need be). Are there any preferences for profiling tools/output format, or is [ruby-prof](https://ruby-prof.github.io/) & [rails-perftest](https://github.com/rails/rails-perftest/tree/master) ok?
Here's the raw output:
[performanceUnmodified.zip](https://github.com/user-attachments/files/18994103/performanceUnmodified.zip)
[performanceGroupSearch.zip](https://github.com/user-attachments/files/18994115/performanceGroupSearch.zip)

**Comment by Sindrake** [comment]
I don't have a preference for profiling, and to be perfectly honest I am not terribly experienced with it in ruby myself.

I once again apologize for not following up on the project. I've been a little busy with the UI side of things.
I appreciate all the work you've been putting into it.

What is the status of this project?
I would like to have it tested, reviewed, and merged sooner rather than later.
Although due to its size, I would probably need to ask other devs on the staff team for help with that.

**Comment by jmortiger** [comment]
Thus far, everything works exactly as intended; barring unacceptably poor performance or a disagreement about implementation, the only thing that would need to be done is removing a few constant variables I've placed to allow for easy behavior tweaking during review (e.g. the `TagQuery::SETTINGS` hash, `ElasticQueryBuilder::GLOBAL_DELETED_FILTER`, & `ElasticQueryBuilder::ERROR_ON_DEPTH_EXCEEDED`). Other than that, I believe everything's production ready.

I'm currently rolling back some of the unnecessary Rubocop-appeasing changes I made to slim the diff & minimize the review burden (some of them may also have ~likely slight~ [performance implications](https://github.com/fastruby/fast-ruby), but they can be reintroduced if needed after the root implementation is deemed strong). I'll check after I'm done slimming and then sanity checking, but it should only be the 12 source files I listed up top, + my regenerated `.rubocop_todo.yml`.

**Comment by jmortiger** [comment]
Actually, considering post SQL searches are already 2nd class searches & don't support some metatags (e.g. `inpool` & `pending_replacements`), I think scrapping grouped searches for stuff like comments & notes altogether would probably be fine, which would result in 11 source files + my regenerated `.rubocop_todo.yml`. As stated at the top, simply removing the changes to `app/logical/post_query_builder.rb` and adding `can_have_groups: false` to the instantiation at the top of `PostQueryBuilder.search` is all that's needed to change this.

**Comment by jmortiger** [comment]
With the current tokenizer, quoted metatags don't need to be separated from the token that follows them by whitespace, a behavior I kept for the new tokenizer, and extended so groups do the same. While this kinda makes sense from a user's perspective and is ~minimally~ helpful, this complicates simple scans of a query string. 

For example, to circumvent the overhead of a group search, a quick scan of the query string checks if there are any groups in the query string. In the best case scenario, you could directly find and discard quoted metatags (as they could contain group delimiters inside) and then search the remaining text for groups, which is fairly simple and performant; however, because quoted metatags can be preceded by syntactically valid groups & other quoted metatags as well as the start of a string and non-quoted whitespace, groups must also be identified; but, because groups are invalid within quoted metatags (& can be preceded by syntactically valid quoted metatags & other groups as well as the start of a string and non-quoted whitespace), quoted metatags would have to be identified first.

This catch-22 means the only way to correctly match all groups and quoted metatags is to match them one by one from the start of the string. This may still be faster in some cases, but both adds a lot of bloat and reduces the potential of simple regex match checks. As I hadn't yet realized this, `TagQuery.scan_metatags` does fail to match metatags preceded by a group without separating whitespace.

Considering this is likely an unintended side-effect of the original `TagQuery.scan` implementation (this behavior isn't checked for in any preexisting unit tests from what I can tell) and, from the user's perspective, is at best a minor convenience and at worst a perplexing inconsistency, I'll be changing this behavior in the tokenizer and optimizing for it; both groups and quoted metatags must be followed by a space, otherwise, they will be matched as a standard tag. Is this acceptable, or should I maintain the prior behavior and fix the corner cases to support this?

Tl;dr: is it ok to force all search tokens to be separated by whitespace for the sake of code clarity, brevity, performance, and maintainability? E.g. `quoted:"metatag with spaces"another:"quoted metatag"` is tokenized as `[ 'quoted:"metatag', 'with spaces"another:"quoted', 'metatag"' ]` instead of `[ 'quoted:"metatag with spaces"', 'another:"quoted metatag"' ]`

**Comment by Sindrake** [comment]
That's one hell of an edge case, yeah.

I think that it's fine to give gibberish results in response to a gibberish query.
If the user does not put spaces between the tags, it's their problem.

**Comment by jmortiger** [comment]
The new tokenizer and its associated improvements is complete. Next thing is probably revising the test suite to catch any other edge cases I missed, then maybe another profiling pass?

**Comment by Sindrake** [comment]
Sounds good to me!

**Comment by Sindrake** [comment]
Sorry, I keep creating more merge conflicts by editing the rubocop TODO.

**Comment by jmortiger** [comment]
It's all good, I prefer that to letting it grow by keeping dead RuboCop rules in there. It's a good reminder to regen before merging anyways.

**Comment by jmortiger** [comment]
Alright, aside from testing and picking an ideal `has_groups?` implementation (& of course expanding the test suite), I don't see much more that can be done. Tomorrow I'll probably just pick one and pull the rest into a gist in case they're needed. Other than that and some nice-to-haves, coming up with a few more robust integrations tests for `post_test.rb` is about all I can think of. After picking an implementation, I'll try to only touch tests, rubocop, & merges to make it easier in case anyone wants to review.

**Comment by Sindrake** [comment]
Excellent, sounds good to me.
I would like to have this PR reviewed in time for next week's release. I will hopefully have time to do so over the weekend.

**Comment by jmortiger** [comment]
I'll only be pushing merges, doc comments & expansions to the test suite (specifically `test/unit/tag_query_test.rb`) for the next few days unless I find a bug through that, in which case I'll make note of such in a comment here. Let me know if you need me.

**Comment by Sindrake** [comment]
Well, I've been reviewing and testing this for the past 5 hours, and I do not have any complaints.
As with everything, I am sure that there will be issues that I've missed – but those would be more easily found by sending this to be tested by users, rather than attempting to polish it in a development environment forever.

Unless you have any objections, I would prefer to have this PR merged tomorrow.

**Comment by jmortiger** [comment]
No objections at all! The only notes I'd have is
* I'm split on removing the `TagQuery::SETTINGS`, `ElasticPostQueryBuilder::GLOBAL_DELETED_FILTER`, & `ElasticPostQueryBuilder::ERROR_ON_DEPTH_EXCEEDED` constants (assuming the currently selected settings are ok); it might be convenient to leave it to make any future troubleshooting easier & simply trim these settings constants as the code gets battle-tested
* The group search for `PostQueryBuilder` (& consequently the comment, note, upload, etc. searches) is still [slightly bugged with complex nesting](https://github.com/jmortiger/e621ng/commit/c86d9a122799776e4ba414980a1311bb6f4b2c86); imo it's good enough to leave for now (so long as its a known bug), but (since these searches are already less powerful than post search) this could be toggled off by setting `PostQueryBuilder::CAN_HAVE_GROUPS` to false. Are you ok with leaving it enabled, or should it be toggled off for now?
Other than that, I'm perfectly happy to merge it whenever you want :+1:

**Comment by Sindrake** [comment]
I am fine with leaving the settings constants as they are for now.
If no adjustments are needed there, they can be removed down the line.

As for the other matter – I think I would prefer to not give people the expectation that tag groups work outside of the regular post search. Things like comment, note, and upload searches are utilized primarily by power users and staff members – and they would need the search results to be accurate and reliable. Of course, the searches they perform are usually very simple, so the chances of them running into the aforementioned bug are not great, but I still think that it's best to avoid it entirely.

**Comment by jmortiger** [comment]
That makes sense; I've toggled off that flag.

**Comment by jmortiger** [comment]
I'll be leaving this as is for now to remove the need for extra testing.

**Comment by Sindrake** [comment]
Sorry, things got a little hectic with the April Fool's event... and with bugfixes for it.
I've reviewed the PR, and will merge it in time for the Wednesday's release, when it'll be deployed to e926 for testing.

**Comment by jmortiger** [comment]
Glad to hear it! Thanks for taking the time to review & merge it, I'm sure there's never enough to go around.

**Comment by Sindrake** [comment]
Had a test fail randomly in a completely unrelated commit.
```
Failure:
TagQueryTest#test_: Parsing a query: should correctly handle wildcards.  [test/unit/tag_query_test.rb:745]:
Expected: ["acb", "azb"]
  Actual: ["azb", "acb"]

bin/rails test test/unit/tag_query_test.rb:742
```

Might be a one-off issue, though, since it has not happened again.

By the way, do you have Discord, by any chance?
Might be easier to talk there, in case fixes need to be made or some such.
